### ISSUE TITLE
RELAYplate Not Working

### ISSUE THREAD

**Issue by BBQRules** [mharsch/node-red-contrib-pi-plates#5]
I set up a Pi with Buster Lite and a RELAYplate. I have SPI running and python 2.7.

Triggering the relays using the python commands on the RELAYplate website works without issue.

If I add the RELAYplate to Node-Red and use a "on" trigger, nothing happens.

Node-Red log shows:

pi@raspberrypi:~/.node-red $ node-red-log

    RP.relayON(addr, relay)
  File "/usr/local/lib/python2.7/dist-packages/piplates/RELAYplate.py", line 75, in relayON
stderr:     VerifyADDR(addr)
  File "/usr/local/lib/python2.7/dist-packages/piplates/RELAYplate.py", line 175, in VerifyADDR
    assert (relaysPresent[addr]==1),"No RELAYplate found at address "+addr_str
AssertionError: No RELAYplate found at address 0
code: 1 signal: null

**Comment by mharsch** [comment]
The address that the error is referring to is the physical jumpers on the PiPlate (the plate address in the stack).  Are you sure the configuration of the relay node matches the relay plate physical address (configured when you first add a PiPlate node)?

**Comment by BBQRules** [comment]
Thanks for your help with this Mike.

I was wondering the same thing but I'm pretty sure it's correct. The
pi-plate is address 0 out of the box and I didn't change any jumpers.

From the Pi-Plates website I used these commands and it worked fine:

import piplates.RELAYplate as RELAY
>>> RELAY.getID(0)
'Pi-Plate RELAY'
>>> RELAY.relayON(0,3)

I also tried setting up a number of triggers going into a number of
pi-plate addresses with no switching as a result. You'll notice in the
attached screen shot that the message coming from the top inject does not
pass through the Plate0 node.
[image: image.png]

The top relay node config:

[image: image.png]

[image: image.png]

**Comment by mharsch** [comment]
Ok, I've seen this symptom before, but I'd like to try to narrow it down.  The error "AssertionError: No RELAYplate found at address 0" comes from python code running under the hood.  If we can reproduce the problem with just a python program (without running any Node-RED at all) that will help us narrow down the issue.  When I've seen this before, it happens more frequently when the system is under load (causing the python program to possibly be interrupted while it's doing SPI communication with the PiPlates).  When Node-RED is starting up, the pi is under load and so that's why it happens more frequently for us.  Anyway, could I ask you to write a python test program that toggles relays and loops forever, then run some other programs at the same time to put a load on the system?  If I'm right, you'll be able to crash your python program with the same error (eventually).

**Comment by mharsch** [comment]
Ok, I think I see at least part of the problem here.  You can't have plates configured with addresses that do not exist (e.g. a plate configured with address 1 when only a single plate with address 0 is installed).  As soon as the underlying python script gets an incorrect address, it crashes and does not restart until the entire Node-RED service is restarted.  So, the fix here is to remove all the configuration nodes for addresses that are not present, and then restart Node-RED (or reboot the whole pi) and try again.  Let me know if this works.

**Comment by BBQRules** [comment]
Gave it a try deleting the unused nodes and rebooted pi.

Same issue.

I'm working on the toggle test script now.

What would you suggest I run to ta the processor when the script is ready?

**Comment by mharsch** [comment]
Could you make sure that the "configuration nodes" associated with the non-existent plates are also removed?  Check the 'configuration nodes' section in the right-hand pane then double click on any unused nodes and click 'delete'.  Then reboot and re-test.  Thanks.

**Comment by BBQRules** [comment]
Yup. Did that.

Still not working.

Sorry.

**Comment by BBQRules** [comment]
So I retooled a script I got off YouTube. What would you suggest I do to
burden the pi?

#!/usr/bin/python
import piplates.RELAYplate as RELAY
import time

relayList = [1, 2, 3, 4, 5, 6, 7]

for i in relayList:
     RELAY.relayOFF(0,1)

SleepTimeS = 0.05
SleepTimeL = 0.05
# main loop
try:
     while True:
        for i in relayList:
           RELAY.relayOFF(0,i)
           time.sleep(SleepTimeS);
           RELAY.relayON(0,i)
           time.sleep(SleepTimeS);
           RELAY.relayOFF(0,i)
           time.sleep(SleepTimeS);
           RELAY.relayOFF(0,i)
           time.sleep(SleepTimeL);
# End program cleanly with keyboard
except KeyboardInterrupt:
      print " Quit"

**Comment by BBQRules** [comment]
This is my Node-Red setup at this stage:

[image: image.png]

**Comment by mharsch** [comment]
Note: this issue originally presented while running on on a Raspberry Pi Model B.  It can also be reproduced using a Pi Zero (which uses the same CPU).  It's much easier to reproduce this on the slower CPU variants, because the root cause is sensitive to system load and Node-RED is fairly CPU intensive when starting up.  The Node-RED Pi-Plates nodes have an underlying python process that is launched when Node-RED starts up.  This process instantiates python objects that are used to send commands to the plate stack.  At the time of initialization, the relay plate object scans the address range (0-8) for plates of type RELAYplate and keeps a table of discovered devices.  If this scan fails to enumerate a plate (which can easily happen if the system is under load) then the state of the table that tracks relay plates present in the system will be incomplete (and will stay incorrect for the life of the process -- a long running process in this case).  As soon as a command comes in for a plate that the underlying python object failed to initialize, the process crashes with the "No RELAYplate found at address <x>" message.  The fix here needs to be twofold: first we need to fix issue #3 so that the underlying python process can survive unexpected input.  Second we need to get retry logic added upstream to the official Pi Plates code so that undiscovered plates can be picked up and used after they were missed during initialization.

**Comment by PCRover** [comment]
Did this issue get resolved? I appear to be experiencing the same problem with a Pi3 vB.

**Comment by mharsch** [comment]
No, this issue is still unresolved.  If you could provide more details of your specific configuration, that would help.  NodeRED version, pyplates (python package) version, which plates and how many are installed / jumper addresses.  Is the problem intermittent or consistently repeatable for you?

**Comment by mharsch** [comment]
Thanks for sharing.  /cc @pi-plates

**Comment by DougSmythies** [comment]
Mike, thank you for your reply. I see you copied Pi-Plates. Please be aware that I have been working with Pi-Plates support on this the whole time.
